Message Replies

AndrewWay opened the pull request:
Added functionality for allowing repositories to reply to clients which are requesting data they contain.

Commit:
Work in progress for replying to clients making requests to repositories

Commit:
Updated Unix Domain Stream Socket files to pass the file descriptor of the process that made contact with it.
Added code to allow replying to socket connections after data is recieved from a client.

Commit:
Modifying Messaging library for replies

Commit:
Made UnixDomainStreamSocketClient publicly inherit UnixDomainStreamSocket to allow for child classes to access buffer to check for replies

Commit:
Added getter function for reading buffer contents in unix domain stream socket

Commit:
Updated ProcessMessage function with new argument for repositories

Commit:
Reply is returned as string through the LorisMessenger

Commit:
MessageSenderInterface has a single SendDataMessage function which returns a string. Returns empty string if no request keys exist in the message.

Commit:
Added Bool function for indicating whether a DataMessage has request keys
There is now just a single "Send" function in LorisMessenger, regardless of whether you wait for a reply

Commit:
Added constructor in DataMessage for building from string

Commit:
LorisMessenger reads from socket

Commit:
Adding several logging statements. DataMessage being constructed by repository to reply to clients still is not working properly.

Commit:
Fixed bug where random data was being prepended to replies from repository

Commit:
Fixed bug in handling replies from repository. Replies work now.

s-axford commented on core/src/repositories/cc/PayloadRepository.cc:
```diff
@@ -23,7 +23,7 @@ int PayloadRepository::AddKeysToWatchList(){
 
 
-int PayloadRepository::ProcessMessage(DataMessage message){
+int PayloadRepository::ProcessMessage(DataMessage message,int client_file_descriptor){
```
I notice this added parameter is not being used in most Repository's. Is that going to change?

liamhebert replied:
Additionally, is this parameter something that should be attached to the message? It seems like this could be contained in `DataMessage`

AndrewWay replied:
I removed this argument in the latest commit

s-axford commented on core/src/repositories/cc/PowerRepository.cc:
```diff
@@ -26,7 +27,7 @@ int PowerRepository::AddKeysToWatchList(){
 }
 
-int PowerRepository::ProcessMessage(DataMessage message){
+int PowerRepository::ProcessMessage(DataMessage message,int client_file_descriptor){
```
Nothing in this method is PowerRepository specific. Are these going to have different implementations for different Repositories? If not we may be able to have a central implementation in the Repository Class.

liamhebert replied:
I agree with this. If you are planning on adding specific handling for each subsystem in these repos classes 
then I think that those would be better suited for another class. Adding both handling and processing in these repositories classes would violate the [Single Responsibility Principal](https://en.wikipedia.org/wiki/Single_responsibility_principle)

AndrewWay replied:
Agreed. I think within "ProcessMessage", we can perform additional processing aside from extracting incoming data and replying with requested data. In which case, we can create some other class which is responsible for the processing

s-axford commented on core/src/repositories/header/Repository.h:
```diff
@@ -8,18 +8,24 @@
     std::vector<Key> watch_list_;
 
-    int ReturnData(DataMessage request_message,DataMessage& return_message);
+	Identifiers identifiers_;
```
If we are storing the individual identifier do we also need to store this struct with ALL the identifiers? Can we just keep one (The struct OR the individual identifier)?

liamhebert replied:
If we are continuing with the pattern of having sub-repositories then each child repository should store the single identifier that it requires. The identifier field would then be declared `virtual` in this class

AndrewWay replied:
Will delete the Identifier struct

s-axford commented on internal-messaging/message/cc/DataMessage.cc:
```diff
@@ -88,7 +108,10 @@ void DataMessage::Flatten(char* msg) {
     strcat(msg, "|");
-
+    /*
+    *	TODO for Spencer: Can you put these delimiting string values in constants somewhere?
```
Okay

s-axford commented on internal-messaging/unixdomain/cc/UnixDomainStreamSocketClient.cc:
```diff
@@ -35,22 +35,29 @@ int UnixDomainStreamSocketClient::Send(char message[]) {
-
-int UnixDomainStreamSocketClient::SendMessageAwaitReply(char message[]) {
+//TODO Should I return a string or should I pass in a char array pointer...
+string UnixDomainStreamSocketClient::SendMessageAwaitReply(char message[]) {
```
This method works as C code, but changing it to rely on the C++ string means it will no longer compile to raw C. Not sure if that is a concern or not.

liamhebert replied:
I don't think any of the sub systems are based on raw C code so I think it would be safe to switch this to a C++ string

AndrewWay replied:
As per mine and Spencer's conversation, we will have a code layer built from C which handles raw processing of messages, and a C++ layer on top which uses strings.

s-axford commented on internal-messaging/unixdomain/cc/UnixDomainStreamSocketServer.cc:
```diff
@@ -55,13 +55,24 @@ void UnixDomainStreamSocketServer::WaitForConnection() {
+        	/*
+        	 * TODO: I think there is an issue with how we are calling ReadFromSocket.
+        	 * Why is 255 being used as a capacity? What happens if the message is over 255?
+        	 * What happens if its under 255? -Andrew
```
Good catch. Perhaps we can make this a variable that can be different for each Repository. (I think it should be constant for each repository at the least as if we are always changing it there is a huge opening for mistakes)

s-axford commented on internal-messaging/unixdomain/header/UnixDomainStreamSocketServer.h:
```diff
@@ -22,7 +22,7 @@ using std::string;
 */
 //REF: http://www.linuxhowtos.org/C_C++/socket.htm
-class UnixDomainStreamSocketServer : protected UnixDomainStreamSocket {
+class UnixDomainStreamSocketServer : public UnixDomainStreamSocket {
```
Why are we removing UnixDomainStreamSocketServer permissions to access protected members of UnixDomainStreamSocket?

AndrewWay replied:
returned it to protected

liamhebert commented on core/src/repositories/cc/Repository.cc:
```diff
@@ -6,15 +6,24 @@
+    char msg[255] = "";
+	message.Flatten(msg);
+	cout << "Replying to client at " << client_file_descriptor << " with message: " << msg << endl;
+	WriteToSocket(msg,client_file_descriptor);
```
I'm a little confused by the purpose of this method. Is this meant to be a message received confirmation of sorts? Or would this be how we would receive the results of a query?

liamhebert replied:
It may be worth adding a comment here that describes how this method should be used

AndrewWay replied:
Added comment to the header file

liamhebert commented on core/src/repositories/cc/Repository.cc:
```diff
@@ -6,15 +6,24 @@
 #include <string>
 
 Repository::Repository(std::string socket_path)
```
Taking a look through this class. I think many of the message sending functions should be relegated to a single interface instead of being combined with data storage. Similar to the above, this appears to violate the [Single Responsibility principal](https://en.wikipedia.org/wiki/Single_responsibility_principle)

AndrewWay replied:
Hmm I may have to get Repository to use a UnixDomainSocketClient rather than inherit from it.

s-axford replied:
We could keep repository as the class where messages come in and get processed, then have another class that handles the replys? That would keep the purpose of repository as processing each message, And if a reply is needed we ship it off to another class? Thoughts?

AndrewWay replied:
I'm going to revisit this through an architectural perspective

liamhebert commented on core/src/repositories/cc/Repository.cc:
```diff
@@ -6,15 +6,24 @@
+}
+
 //Check if the key is contained in the watch list for the repository.
 bool Repository::WatchListContainsKey(unsigned int key){
```
Unrelated to this PR but `watch_list_` should be stored in a map rather then a vector. That would remove the necessity of this method

AndrewWay replied:
Put it as a TODO

liamhebert commented on emulation/src/main.cc:
```diff
@@ -7,6 +7,7 @@
 #include <cstdio>
+#include <string>
 using std::cout;
 using std::endl;
```
Unrelated to this PR but I feel like this should be driven by a testing framework such as [Google Test](https://github.com/google/googletest)

AndrewWay replied:
ill look this up

liamhebert commented on internal-messaging/message/cc/DataMessage.cc:
```diff
@@ -15,44 +15,64 @@ DataMessage::DataMessage(unsigned int sender, unsigned int recipient, long time,
+ * This function assumes that flat is the "flattened" version of a pre-existing DataMessage
+ * TODO for andrew: give this a better name
+ */
+int DataMessage::BuildFromCharacters(char * flat){
```
`DataMessage::AssembleMessage`? Descriptive names are always best so either works

liamhebert replied:
Additionally, there are many iterations and verifications here which I feel should have been handled by the class that assembles the messages. Ideally, the data should be enforced to be a specific format with proper enter and exit conditions without ambiguity

AndrewWay replied:
Will revisit in architecture

liamhebert commented on internal-messaging/message/cc/DataMessage.cc:
```diff
@@ -15,44 +15,64 @@ DataMessage::DataMessage(unsigned int sender, unsigned int recipient, long time,
+	            i++;
+	        }
+	        try {
+	        	std::cout << "Adding: " << hex_string << "to requests in DataMessage" << std::endl;
```
This indentation is weird

AndrewWay replied:
copy and paste + laziness shenanigans

liamhebert commented on internal-messaging/message/cc/DataMessage.cc:
```diff
@@ -15,44 +15,64 @@ DataMessage::DataMessage(unsigned int sender, unsigned int recipient, long time,
+	    while(flat[i] != '|' && flat[i-1] != '|'){
+	        // Get next request
+	    	//TODO fix bug where infinite loop is reached due to no requests in the message
+	        while(flat[i] != '-' && flat[i] != '|'){
```
This seems oddly redundant to have two `while` loops with the same condition (`flat[i] != '|'` & `flat[i-1] != '|'`). There should be a comment here explaining what `|` and `-` are in this loop. Maybe instead of passing a raw char pointer, a c++ string should be used instead which holds the size of the string. This would transform this loop into being something like `while(flat[i] != '-' && i < flat.size())`

AndrewWay replied:
@s-axford ?

liamhebert commented on internal-messaging/messaging-service/cc/MessageSenderInterface.cc:
```diff
@@ -12,13 +12,27 @@ void MessageSenderInterface::SetRecipient(unsigned int recipient){
 
-void MessageSenderInterface::SendDataMessage(DataMessage message) {
+//Send message to socket and await reply
+string MessageSenderInterface::SendFlattenedMessageAwaitReply(char message[]) {
```
This seems dangerous due to the [Two General's Problem](https://en.wikipedia.org/wiki/Two_Generals%27_Problem). There should be a limit to how long this waits for (this might be implemented in `SendMessageAwaitReply`, just haven't seen it yet)

liamhebert replied:
Additionally, sending a message like this should include a flag that indicates that the OS is awaiting a reply.

AndrewWay replied:
Reply flag added as new feature to be implemented

liamhebert commented on internal-messaging/messenger/header/LorisMessenger.h:
```diff
@@ -25,7 +25,9 @@ class LorisMessenger{
 
 	//Method to send stored message to recipient from sender (Using Identifer int - see Identifier.h for correct values)
-	int Send(unsigned int recipient, unsigned int sender);
+	//TODO implement time out in UnixDomainStreamSocketClient for when server doesn't respond with reply
```
Ah seems like you read my mind with the above message. Cool!

liamhebert commented on internal-messaging/unixdomain/cc/UnixDomainStreamSocketClient.cc:
```diff
@@ -35,22 +35,29 @@ int UnixDomainStreamSocketClient::Send(char message[]) {
     if (WriteToSocket(message, socket_file_descriptor_) != 0) {
         error("ERROR SENDING MESSAGE");
-        return 1;
+        return "FAILURE";
```
I feel like the reply should be stored in a separate pointer with this function returning true or false. Returning strings like this leads to unnecessary complexity where we would have to verify if the reply is actually accurate and removes the possibility of sending a failure response

liamhebert commented:
Good stuff Andrew! Quick minor points that is unrelated to my comments, lots of your method headers have odd spacing such as `foo(string bar,int zoo)`. It seems like the standard across your repo to include a space after each comma (its also a little nicer to read). Another point, a quick glace at the testing code includes blocks of commented out code, not the nicest looking thing. Lastly, PR's shouldn't introduce `TODO` comments as they should start and end a feature. Things like this might be better discussed in an issue ticket or via slack. Hope I'm not being too nit picky :)

liamhebert commented:
Last thing, I mentioned this briefly to @s-axford but I think it would be very important (and useful for @DalhousieSpaceSystemsLab/payload ) if changes to the message sending code were cataloged by a [Changelog](https://keepachangelog.com/en/1.0.0/) using [Semantic Versioning](https://semver.org/). This would make it way easier to know if modifications are API breaking and what parts should be modified to accommodate this. Things that I would consider API breaking would be changing the return types and response types (such as this PR) which require re-tooling of the sub-systems that rely on this. If you do realize that a API breaking change is necessary, a quick message in #payload on slack would be much appreciated :)

Commit:
Removed client file descriptor as arguments for functions which handle replies to clients connecting to repositories

Commit:
Added almost empty script for building core project

Commit:
Repository Objects reply to clients without calling functions that take a client file descriptor as argument

Commit:
Made UnixDomainStreamSocketServer's inheritance from UnixDomainStreamSocket protected instead of public

Commit:
Added a couple doxygen compatible comment blocks for Repository

Commit:
Moved code common to all children of repository to the Repository class

Commit:
Removed identifiers as attribute for repository classes
GetIdentifiers() is now pure virtual function repository_identifier() in the Repository class.

Commit:
Pull request clean up

Commit:
UnixDomainStreamSocketClient function SendmessageAwaitReply returns int rather than string
